### ISSUE TITLE
Building shared libraries not working in osx

### ISSUE THREAD

**Issue by modlfo** [ocaml/ocamlbuild#124]
I tried to use the targets `.native.{so,dll,dylib}` to build a shared library but it did not worked. First, the target `.native.dylib` is not valid. Then I tried the target `.native.so` and I got the error

```
clang: error: invalid argument '-bundle' not allowed with '-dynamiclib'
File "caml_startup", line 1:
Error: Error during linking
Command exited with code 2.
```

This can be reproduced as follows, create a hello world (main.ml) and run:
```
$  ocamlbuild main.native.so
```

**Comment by gasche** [comment]
That sounds like an OSX tooling backard-incompatibility issue -- there are some all the times around basic development tools. What are your OCaml, OCamlbuild and OSX versions?

**Comment by modlfo** [comment]
- ocaml 4.03.0
- ocamlbuild 0.9.3
- OSX (Sierra) 0.12.1 
- XCode 8.1

**Comment by dbuenzli** [comment]
Can reproduce here's the full failing invocatio: 

```
> /Users/dbuenzli/.opam/4.03.0/bin/ocamlopt.opt -verbose -output-obj -cclib -shared main.cmx -o main.native.so
+ clang -arch x86_64 -c -o '/var/folders/g4/nl6g_ytj63bbn9d45_jhggxc0000gn/T/camlstartup8da423.o' '/var/folders/g4/nl6g_ytj63bbn9d45_jhggxc0000gn/T/camlstartup11ffe4.s'
+ gcc -bundle -flat_namespace -undefined suppress                    -Wl,-no_compact_unwind -o 'main.native.so'   '-L/Users/dbuenzli/.opam/4.03.0/lib/ocaml'  '/var/folders/g4/nl6g_ytj63bbn9d45_jhggxc0000gn/T/camlstartup8da423.o' 'main.o' '/Users/dbuenzli/.opam/4.03.0/lib/ocaml/stdlib.a' '-shared' '/Users/dbuenzli/.opam/4.03.0/lib/ocaml/libasmrun.a' 
clang: error: invalid argument '-bundle' not allowed with '-dynamiclib'
File "caml_startup", line 1:
Error: Error during linking
```
I don't see any `dynamiclib` argument on this line though bu some search suggest that is due to the `-shared` option on that line.

**Comment by gasche** [comment]
This seems to be the same issue as OCaml's [MPR#6927](https://caml.inria.fr/mantis/view.php?id=6927), for which @whitequark seemed to have a solution but also to be unsure it wouldn't introduce new issues. Could I let people that actually use MacOS ( @dbuenzli ? @whitequark ? ) decide what is best on this one?

**Comment by whitequark** [comment]
@gasche I have reviewed MPR#6927 with some of my added OS X experience. At this point, in my opinion the obvious right course of action is to keep using `.so` but create dylibs. The paragraph from [this comment](https://caml.inria.fr/mantis/view.php?id=6927#c14211) states:

> The differences between bundles and dylibs are rather subtle. At 10.4 and before, you could not dlopen() a dylib--you could only link to it, in fact there was no dlopen()--and you could use the corresponding NS* API to load and unload a bundle. After 10.4, bundles and dylibs are almost entirely equivalent; you can dlopen() and dlclose() both. The only differences are that 1) you cannot directly link a dylib and 2) rpath is not handled in the same way. (I have grepped ld64, XNU and dyld from Apple's opensource repository to independently confirm this information.)

That was a typo; what I meant is:

> 1) you cannot directly link a bundle

(See also [this StackOverflow question](https://stackoverflow.com/questions/2339679/what-are-the-differences-between-so-and-dylib-on-osx)).

The `.so` extension as well as building a dylib is clearly valid for a `.native.so` and `.byte.so`; for building bytecode stubs or plugins, bundles are slightly semantically closer, but the advantage of not having to 1) special-case stubs and plugins 2) special-case Darwin anywhere in the ecosystem, is overwhelming.

The `-force_load` issue is orthogonal, and can be handled in several ways, e.g. as [described](https://caml.inria.fr/mantis/view.php?id=6927#c14212).

**Comment by gasche** [comment]
Thanks for the extra info. Would you (or someone else in this PR) be willing to propose a patch to implement this course of action?

**Comment by whitequark** [comment]
@gasche But there's already one attached to MPR#6927. Do you want me to reformat it as a GPR?

**Comment by gasche** [comment]
Hm, my understanding was that the patch would on the contrary build a bundle rather than a dylib, removing a conflict with an additional `-shared` flag. (But yes, anything you think is good as a GPR would help.)

**Comment by whitequark** [comment]
OK, I'll do it a bit later.

**Comment by whitequark** [comment]
@gasche https://github.com/ocaml/ocaml/pull/988, https://github.com/ocaml/ocamlbuild/pull/136

**Comment by gasche** [comment]
Thanks a lot! I merged #136 on the ocamlbuild side, so the current master and the next release should fix this bug.
